ArdusatSDK.cpp

```diff
@@ -453,266 +257,1212 @@ const char * valuesToCSV(const char *sensorName, unsigned long timestamp, int nu
   return _getOutBuf();
 }
 
-// Define a version of a ToCSV function that uses the default name for that
-// sensor type
-#define _DEFAULT_SENSOR_NAME_TO_CSV_FXN(NAME) \
-  const char * NAME ## ToCSV( NAME ## _t &input) { \
-    char sensor[50]; \
-    strcpy_P(sensor, NAME ## _sensor_name); \
-    return NAME ## ToCSV ( sensor, input ); \
+/*
+ * Internal helper to build the JSON string at the proper point in the output buffer
+ * with the correct values and labels
+ */
+int _writeJSONValue(char *buf, const char *sensor_name, const char *unit, float value) {
+  char num [32];
+  char format_str[80];
+  // inexact estimate on the number of characters the value will take up...
+  if (strlen(sensor_name) + strlen(unit) + 10 + _output_buf_len > OUTPUT_BUF_SIZE) {
+    return -1;
   }
+  dtostrf(value, 4, 2, num);
+  strcpy_P(format_str, json_format);
+  _output_buf_len += sprintf(buf, format_str,
+			     JSON_PREFIX, sensor_name, unit, num,
+			     calculateCheckSum(sensor_name, 1, value), JSON_SUFFIX);
+  return _output_buf_len;
+}
 
-_DEFAULT_SENSOR_NAME_TO_CSV_FXN(acceleration)
-const char * accelerationToCSV(const char *sensorName, acceleration_t & input) {
-  _headerToCSV(&(input.header), sensorName);
-
-  dtostrf(input.x, 2, 3, _getOutBuf() + _output_buf_len);
-  _output_buf_len = strlen(_getOutBuf());
-  _getOutBuf()[_output_buf_len++] = CSV_SEPARATOR;
+/*
+ * Internal helper to build 3 JSON strings back to back since there are so many sensors
+ * that report 3 values
+ */
+const char * _tripleValueToJSON(const char *sensorName, int labelLength, unsigned char unit,
+             char *labelA, float valueA, char *labelB, float valueB, char *labelC, float valueC) {
+  int nameLength = strlen(sensorName);
+  char nameBuf[nameLength + labelLength];
```

Comment:
Add a `+ 1` to handle the null termination byte (technically this could be included in the `labelLength`, but I hate relying on that.

I actually think rather than relying on a `sprintf` compatible format string for `labelA`, etc., it'd be more useful to just have the labels be simple strings that identify an individual value, and then have a default format string (e.g. `%s%s`) that we use for concatenation here. Then we can dynamically allocate the correct length for `nameBuf` without relying on our caller having to worry about how long the label is going to be.